src/compiler/checker.ts

```diff
@@ -16409,6 +16409,9 @@ namespace ts {
             if (suggestion !== undefined) {
                 errorInfo = chainDiagnosticMessages(errorInfo, Diagnostics.Property_0_does_not_exist_on_type_1_Did_you_mean_2, declarationNameToString(propNode), typeToString(containingType), suggestion);
             }
+            else if (isPromiseLike(containingType)) {

```

Comment:
call `getPromisedTypeOfPromise` instead. this will return you the type of the `T` in `Promise<T>`. this should handle unions and custom promises, etc.. if it is `undefined` then it was not a promise.

you also need to check if the property exists on the promisedType. in other words:

```ts
var p: Pomise<number>;
p.toExponential;  // should report error `Property toExponential does not exist on type 'Promise<number>' did you forget `await`?`

var x: Promise<string>;
p.toExponential;  // should not report the `did you forget await` part.
```